plugin/core/windows.py

```diff
@@ -739,3 +756,145 @@ def _format_request(self, direction: str, method: str, request_id: Any) -> str:
 
     def _format_notification(self, direction: str, method: str) -> str:
         return "{} {} {}".format(direction, self._server_name, method)
+
+
+class RemoteLogger(Logger):
+    PORT = 9981
+    DIRECTION_OUTGOING = 1
+    DIRECTION_INCOMING = 2
+    _ws_server = None  # type: Optional[WebsocketServer]
+    _ws_server_thread = None  # type: Optional[threading.Thread]
+
+    def __init__(self, manager: WindowManager, server_name: str) -> None:
+        self._manager = ref(manager)
+        self._server_name = server_name
+        if not RemoteLogger._ws_server:
+            try:
+                RemoteLogger._ws_server = WebsocketServer(self.PORT)
+                RemoteLogger._ws_server.set_fn_new_client(self._on_new_client)
+                RemoteLogger._ws_server.set_fn_client_left(self._on_client_left)
+                RemoteLogger._ws_server.set_fn_message_received(self._on_message_received)
+            except OSError as ex:
+                if ex.errno == 48:  # Address already in use
+                    debug('WebsocketServer not started - address already in use')
+                    RemoteLogger._ws_server = None
+                else:
+                    raise ex
+        self._start_server()
+
+    def _start_server(self) -> None:
+        if RemoteLogger._ws_server:
+            def start_async() -> None:
+                if RemoteLogger._ws_server:
+                    RemoteLogger._ws_server.run_forever()
+            RemoteLogger._ws_server_thread = threading.Thread(target=start_async)
+            RemoteLogger._ws_server_thread.start()
```

Comment:
I am allergic to threads. This seems to be starting a thread for every logger, in other words for every session. Is it not the intention that there is only one thread running for the entire application instead?

Reply:
`_ws_server` is a (static) class property so it's only initialized once.

Reply:
But `_start_server` is always called in `__init__` of `RemoteLogger`? Should that not be behind an if-condition?

Reply:
The `if` condition could be there but `if RemoteLogger._ws_server:` check here makes sure it works like that. I can move the check there, I agree it makes sense.

Reply:
Done